usb: device: cdc_acm: Use ZLP to detect initial host read

tmon-nordic opened the pull request:
Queue characters in TX ring buffer regardless of interface state. Prevent ECHO on Linux by arming IN endpoint with ZLP when interface is configured and making sure that actual payload is only sent after initialization timeout. The ZLP is not visible to host side applications because the applications are really accessing tty buffer and received ZLP does not modify tty buffer in any way.

See #64024 and #54705 for related discussions.

Fixes #64023

tmon-nordic committed:
usb: device: cdc_acm: Use ZLP to detect initial host read

Prevent ECHO on Linux by arming IN endpoint with ZLP when interface is
configured and making sure that actual payload is only sent after
initialization timeout. The ZLP is not visible to host side applications
because the applications are really accessing tty buffer and received
ZLP does not modify tty buffer in any way.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

gmarull reviewed (dismissed):
Count my +1 as an alternative to https://github.com/zephyrproject-rtos/zephyr/pull/64024, as I do not have enough knowledge to review this properly

carlescufi reviewed (dismissed):
I tested this Pull Request with an nRF52840 Dongle and the following Hosts:

- Linux: Ubuntu and `minicom`, Ubuntu and `cat /dev/ttyACM0`
- Windows 11: Putty
- macOS: `minicom`

In all cases this Pull Request behaved in the manner that a user would expect: the text below appeared upon opening the terminal:
```
*** Booting Zephyr OS build zephyr-v3.5.0-810-gf320200545a7 ***
Hello World! nrf52840dongle_nrf52840
```

carlescufi commented:
cc @xyzzy42

xyzzy42 commented:
The ZLP is a clever workaround for the POSIX race.  Really should be fixed in POSIX too.  E.g., export the termios settings in sysfs so they can be changed before the device is opened.

Does this change the current behavior w.r.t. buffering?

@carlescufi brought it up, https://github.com/zephyrproject-rtos/zephyr/pull/64024#issuecomment-1774854864  The existing behavior is unusual.  If the UART polling API is used, there is no buffering and data is discarded.  But if the UART interrupt driven API is used, then data is buffered indefinitely and this buffering can not be disabled.

@tmon-nordic Then said this, about the buffering, "I would prefer to have CDC ACM not lose any characters unless absolutely necessary (i.e. host is not reading the data and **user disabled hardware flow control**)."

Has it become possible to control the hardware flow control inherent in the CDC ACM protocol?  Last I checked, in #53319, it wasn't possible to do this and the various parties didn't seem to care and/or understand the problem.  Keep in mind that RTS/CTS flow control is unrelated and has no effect on what happens with USB CDC ACM when the host does not read the data.

Because if CDC ACM does not lose any character unless the user disabled hardware flow control, and USB CDC ACM hardware flow control **can not be disabled**, then doesn't that result in CDC ACM never losing any characters ever?

Now imagine a simple example of a program the prints something in a loop while it blinks an LED.  This a ubiquitous "my first LED" sample Arduino program.   If the printf output goes to a UART, then it blinks with nothing connected to the serial port.  But if the output goes to a USB CDC ACM device, then the LED stops blinking as the loop will be blocked on the mandatory USB CDC ACM hardware flow control.

If you've changed it so that the polling API now blocks until a device is connected, it will create the problem that logging to USB CDC ACM blocks with not connected to a host.  And if the interrupt driven API continues to have mandatory flow control, you're going to trade the "missing initial console output" bug reports for "program hangs when not connected to USB host" bug reports.

tmon-nordic commented:
> The ZLP is a clever workaround for the POSIX race. Really should be fixed in POSIX too. E.g., export the termios settings in sysfs so they can be changed before the device is opened.

Discussing Linux changes here is pretty much pointless because this is not the correct place to do so. If you believe this can be changed in Linux (I doubt), then send the PATCH to linux-serial mailing list.

> Does this change the current behavior w.r.t. buffering?

Slightly. This PR removes the "drop because interface is not configured or suspended". Drop on suspend was completely ridiculous (the real solution to that would be issuing wakeup request, which is what Linux expects from CDC ACM device when it claims to support remote wakeup; but this is separate issue and I believe it is outside the scope of this PR). Drop when not configured can be somewhat argued for. However, in my opinion it is better to simply lose the "last chracters" instead of doing shenanigans related to which character should be lost.

> @carlescufi brought it up, [#64024 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/64024#issuecomment-1774854864) The existing behavior is unusual. If the UART polling API is used, there is no buffering and data is discarded. But if the UART interrupt driven API is used, then data is buffered indefinitely and this buffering can not be disabled.
> 
> @tmon-nordic Then said this, about the buffering, "I would prefer to have CDC ACM not lose any characters unless absolutely necessary (i.e. host is not reading the data and **user disabled hardware flow control**)."
> 
> Has it become possible to control the hardware flow control inherent in the CDC ACM protocol?

Yes, it is now possible since c152e0980cf6ae45a1922c7ba2a59ebdcc419664. This change applies on top of the "usb: device: cdc_acm: block in uart_poll_out() routine" and essentially implements "not lose any characters unless absolutely necessary".

> Because if CDC ACM does not lose any character unless the user disabled hardware flow control, and USB CDC ACM hardware flow control **can not be disabled**, then doesn't that result in CDC ACM never losing any characters ever?

Disable `hw-flow-control` in device tree for the CDC ACM instance in question and it'll drop characters if TX buffer is full.

> Now imagine a simple example of a program the prints something in a loop while it blinks an LED. This a ubiquitous "my first LED" sample Arduino program. If the printf output goes to a UART, then it blinks with nothing connected to the serial port. But if the output goes to a USB CDC ACM device, then the LED stops blinking as the loop will be blocked on the mandatory USB CDC ACM hardware flow control.
> 
> If you've changed it so that the polling API now blocks until a device is connected, it will create the problem that logging to USB CDC ACM blocks with not connected to a host. And if the interrupt driven API continues to have mandatory flow control, you're going to trade the "missing initial console output" bug reports for "program hangs when not connected to USB host" bug reports.

This is up to user to decide now. If the user chooses to go with default it will block, if the user sets `hw-flow-control` to `false` then it'll just lose the "middle characters" (i.e. the initial ones that made it to buffer will be sent to host once it starts reading data, the "middle ones" that didn't make it won't be there, and the following characters will make it just fine if there is space for them in TX buffer).

xyzzy42 commented:
> If you believe this can be changed in Linux (I doubt)

I've made hundreds of changes to Linux.  In my experience it is Zephyr which is impossible to change and making PRs is a complete waste of effort.  I guess who you work for provides a different perspective.

> > Has it become possible to control the hardware flow control inherent in the CDC ACM protocol?
> 
> Yes, it is now possible since [c152e09](https://github.com/zephyrproject-rtos/zephyr/commit/c152e0980cf6ae45a1922c7ba2a59ebdcc419664). This change applies on top of the "usb: device: cdc_acm: block in uart_poll_out() routine" and essentially implements "not lose any characters unless absolutely necessary".

That is amazing.  Given the reception when I reported this a year ago, it seemed that jfischer could barely contain his contempt for how stupid he though the idea that blocking forever on output to an unconnected USB device might be undesirable.

But that change only affects the polling API.  The interrupt driven API maintains the mandatory flow control blocking behavior that can not be disabled.

Is there any reason why API choice should make a difference in if flow control can be disabled or not?

There will be a few other effects from this commit.

The cost of logging to an unconnected USB port will increase as the output won't be discarded early before entering the ring buffer.

When the ring buffer fillls up, which it didn't previously do since the data was discarded early, it will trigger a warning or info level log messages (the levels are different for polling vs interrupt API) every time data is discarded because it can not enter the full ring buffer.  These extra log messages are possibly significant more overhead for output to unconnected USB CDC ACM devices.  Far worse, if those log messages are then in turn are sent to an unconnected USB CDC ACM device as a log backend, then I do not anything will stop the log loop.

tmon-nordic commented:
> > If you believe this can be changed in Linux (I doubt)
> 
> I've made hundreds of changes to Linux. In my experience it is Zephyr which is impossible to change and making PRs is a complete waste of effort. I guess who you work for provides a different perspective.

Please CC me when you send the patch to linux-serial.

> But that change only affects the polling API. The interrupt driven API maintains the mandatory flow control blocking behavior that can not be disabled.
> 
> Is there any reason why API choice should make a difference in if flow control can be disabled or not?

I would prefer that this gets off-scope from this PR, but what would be the good approach then? Should the interrupt API tx ready simply be tied to "there is space in tx ringbuf"?

> The cost of logging to an unconnected USB port will increase as the output won't be discarded early before entering the ring buffer.

Does it really matter? If you want to log to CDC ACM, then why wouldn't you want to buffer the data if possible?

> When the ring buffer fillls up, which it didn't previously do since the data was discarded early, it will trigger a warning or info level log messages (the levels are different for polling vs interrupt API) every time data is discarded because it can not enter the full ring buffer. These extra log messages are possibly significant more overhead for output to unconnected USB CDC ACM devices.

What would be the solution then? Change the `LOG_INF()` to `LOG_DBG()` or remove it all-together?

> Far worse, if those log messages are then in turn are sent to an unconnected USB CDC ACM device as a log backend, then I do not anything will stop the log loop.

The loop won't happen because CDC ACM class logging is forced off at compile time if it detects the potential for this problem:
https://github.com/zephyrproject-rtos/zephyr/blob/f320200545a7226c632cc3f733cffbbd5d495931/subsys/usb/device/class/cdc_acm.c#L58-L67

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -223,12 +223,27 @@ static void cdc_acm_write_cb(uint8_t ep, int size, void *priv)
 	}
 
-	k_work_schedule_for_queue(&USB_WORK_Q, &dev_data->tx_work, K_NO_WAIT);
+	/* If size is 0, it means that host started polling IN data because it
+	 * has read the ZLP we armed when interface was configured. This ZLP is
+	 * probably the best indication that host has started to read the data.
+	 * Wait initialization timeout before sending actual payload to make it
+	 * possible for application to disable ECHO. The echo is long known
+	 * problem related to the fact that POSIX defaults to ECHO ON and thus
+	 * every application that opens tty device (on Linux) will have ECHO
+	 * enabled in the short window between open() and ioctl() that disables
+	 * the echo (if application wishes to disable the echo).
```
It only works once, right?
This is all just in case CDC ACM is used for protocol-less streaming of something like logging (everyone must understand it). And for a sample, not static like hello_world, but logging something periodically, if the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal. This is probably okay, but IMHO it needs to be documented so we do not have to triage it again.

Also, people have started to think that it is fine not to wait for DTR because this PR will magically fix it, which is obviously not the case. NACKing IN tokens for a period of time and not adding anything to the UART FIFO until DTR are two different things. (e.g. PR #64039, worse, this board now enables a subsystem using an other SYS_INIT, which is not just silly and not necessary but also against the board porting guide). Therefore the commit message should clearly state that it does not release the user from waiting for DTR.

xyzzy42 replied:
> the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal.

No, that won't happen.  I explained this in one of the earlier PRs that has been linked.  Because the termios settings persist across opens, the race of ECHO being on is only present for the first open.  After that it never be turned on again.

But if the device is removed and then re-connected from USB, without resetting it, then on the reconnect the ZLP delay workaround will not re-trigger the problem will be back.

tmon-nordic replied:
> It only works once, right?

Once per configuration which matches tty device lifetime, which includes termios settings.

> But if the device is removed and then re-connected from USB, without resetting it, then on the reconnect the ZLP delay workaround will not re-trigger the problem will be back.

ZLP does retrigger every time device is configured after it the bus has been reset (which always happens on USB device connection) and therefore the problem won't be back.

nordicjm replied:
> No, that won't happen. I explained this in one of the earlier PRs that has been linked. Because the termios settings persist across opens, the race of ECHO being on is only present for the first open. After that it never be turned on again.

Not talking about echo here but I know for sure of one serial port library where that is completely false because I was getting garbage data whenever the port was opened (the port would be at 9600 baud when connected, the program would open it, configure it, set it to 115200 baud but the garbage data from 9600 baud would be emitted. After closing the port, it would revert the settings back to how they were before it opened i.e. 9600 baud, then opening it would give you garbage again), so this is entirely down to applications, just because e.g. picocom does it one way does not mean every tool does it that way

tmon-nordic replied:
> After closing the port, it would revert the settings back to how they were before it opened i.e. 9600 baud, then opening it would give you garbage again), so this is entirely down to applications, just because e.g. picocom does it one way does not mean every tool does it that way

The case where the echo was enabled by any of the applications that did open the port is what we cannot detect in any way. But I believe this is a case that should entirely be treated as host user-space issue, because someone enabled echo on port that shouldn't really have it enabled.

Similarly, there is case where the first application opening tty device does not disable the echo. But in this case this is likely that host user-space really expects the ehco to be on.

nordicjm replied:
> But I believe this is a case that should entirely be treated as host user-space issue

Agreed, can't accommodate every possible way that every serial library uses, accommodating to the sensible/widest used ones is good enough.

xyzzy42 replied:
> Not talking about echo here but I know for sure of one serial port library where that is completely false because I was getting garbage data whenever the port was opened

But a serial port library isn't part of POSIX.  termios persisting across open() is part of the POSIX spec, and thus how the Linux kernel works.  And so this race (ECHO here, but baud rate also races on a RS-232 port) can't be fixed by just writing userspace software that works.

A userspace lib setting the port to the wrong baud rate before changing it to the correct one is clearly a library bug, easily fixable.  And if it's saving/restoring, then the baud can be set with `stty` before it starts so it restores the correct value.

I never meant to say in my earlier comment that writing userspace software that does something broken with termios settings was impossible.

The ZLP here isn't really a true fix, it mitigates the race by adding a delay.  There's an unknown delay from when the port is opened to when ioctl() is called to fix the termios settings.  There's a delay from the ZLP being sent, at open(), and Zephyr CDC ACM sending data (configurable value added in this PR).  I think I recently read someone saying that you can't fix a race by juggling delays.  The real fix is a change in sequencing, but it's on the POSIX/Linux side that it needs to be done.  But even though it's just a mitigation of the race and not a fix, it's clearly a huge improvement worth doing IMHO.

jfischer-no replied:
> > the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal.
> 
> No, that won't happen. I explained this in one of the earlier PRs that has been linked. Because the termios settings persist across opens, the race of ECHO being on is only present for the first open. After that it never be turned on again.

It could happen, because application may restore tty settings or reset to default by something else.

xyzzy42 replied:
> It could happen, because application may restore tty settings or reset to default by something else.

Configure the port disable echo via automatically when it's first detected via a udev rule.  Then the application doesn't have to do it and an application that restores settings will not change them back to incorrect values.

The application could also be full of bugs and crash randomly too.

jfischer-no replied:
> > It could happen, because application may restore tty settings or reset to default by something else.
> 
> Configure the port disable echo via automatically when it's first detected via a udev rule. Then the application doesn't have to do it and an application that restores settings will not change them back to incorrect values.
> 
> The application could also be full of bugs and crash randomly too.

Yes, discussed internally today, and then user does not need this patch.

tmon-nordic replied:
> Yes, discussed internally today, and then user does not need this patch.

The user does need this patch, because no matter how you set up the udev rule there'll be time window between device being `open()` and the `ioctl()` that disables the echo. This time window is what the ZLP tries to mitigate.

xyzzy42 replied:
Yes, exactly.  udev is userspace and can not solve the POSIX race of opening the port and starting the ECHO process before making the ioctl() call on the open port to turn it off.

If a userspace app has silly behavior, this can always be fixed in userspace.  One could go as far as LD_PRELOADing a wrapper that modifies the C lib `tcsetattr()` function to inject alternate termios settings.

xyzzy42 commented:
> > But that change only affects the polling API. The interrupt driven API maintains the mandatory flow control blocking behavior that can not be disabled.
> > Is there any reason why API choice should make a difference in if flow control can be disabled or not?
> 
> I would prefer that this gets off-scope from this PR, but what would be the good approach then? Should the interrupt API tx ready simply be tied to "there is space in tx ringbuf"?

This PR changes the the interrupt driven code path to not discard on fifo fill.  One reason (provide by you and I agree with the reasoning) that this is acceptable is that this can be configured via flow control.  Except this reason does not apply to the interrupt driven code path.  So it doesn't seem off-scope for the PR.  Maybe if the PR didn't touch the interrupt-driven path, but it does.

And yes, I think I would tie tx ready to space in the tx ringbuf.  Haven't tested that to be sure.  It should block (not signal tx ready) when flow-control is enabled and the ringbuf is full, but continue to signal tx ready if flow control is not enabled and discard data on attempts to fill the fifo.  It would achieve the same result as the polling API.  Which is basically the same a real uart would do, other than what data is discarded would be different.  And before anyone says a real uart doesn't discard before it buffers data, what I mean is that a real uart "discards" the data because it sends on a serial line that nothing is listening on.  USB can't do this, so discarding at the ring buffer is the emulation of that behavior.

> > The cost of logging to an unconnected USB port will increase as the output won't be discarded early before entering the ring buffer.
> 
> Does it really matter? If you want to log to CDC ACM, then why wouldn't you want to buffer the data if possible?

The cost of putting data into the ring buffer is likely not significant, but it does exist.  It's a common use case to enable logging, but not actually have anything on the other end of the transport (UART, CDC ACM, RTT, etc.) to receive the logs.  I see it even on production firmware.  We don't need to buffer the logs, because in the case we actually do want them, something would be there to receive the data without additional buffering.  But mostly nothing is there to receive them and we would like them to be discarded as efficiently as possible.  Why not turn off logging?  That's obviously the most efficient.  But then you need a way to turn logging back on.

> > When the ring buffer fillls up, which it didn't previously do since the data was discarded early, it will trigger a warning or info level log messages (the levels are different for polling vs interrupt API) every time data is discarded because it can not enter the full ring buffer. These extra log messages are possibly significant more overhead for output to unconnected USB CDC ACM devices.
> 
> What would be the solution then? Change the `LOG_INF()` to `LOG_DBG()` or remove it all-together?

I'd remove it.  Or have it trigger only once.  Both of these entirely fix the log loop problem, w.r.t. those logging statements, even if log output to CDC ACM is used.
 
> > Far worse, if those log messages are then in turn are sent to an unconnected USB CDC ACM device as a log backend, then I do not anything will stop the log loop.
> 
> The loop won't happen because CDC ACM class logging is forced off at compile time if it detects the potential for this problem:

That commit only disables logging if CDC ACM is the console.  If a logging uart backend is configure to use a CDC ACM device, then I don't see how the commit has any effect at all.  And if the console is a CDC ACM device, but logging is not configured to go to the console (i.e., it instead goes to a logging uart backend or to the RTT backend or elsewhere), then CDC ACM logging is disabled when it serves no purpose.

So the point remains, this PR will cause anything logging to a uart backend that is on CDC ACM to enter an infinite log loop when a host is not connected to empty the CDC ACM ring buffer.

tmon-nordic commented:
> Except this reason does not apply to the interrupt driven code path. So it doesn't seem off-scope for the PR. Maybe if the PR didn't touch the interrupt-driven path, but it does.

To get this off-scope for the PR all that needs to be done is just to keep `cdc_acm_fifo_fill()` unchanged? The ZLP sent in `USB_DC_CONFIGURED` handler will only shift when the first tx callback gets called (from "configured" to "host started polling IN endpoint"), but the overall behavior should be untouched. Then the interrupt driven path can be changed later.

> The cost of putting data into the ring buffer is likely not significant, but it does exist. It's a common use case to enable logging, but not actually have anything on the other end of the transport (UART, CDC ACM, RTT, etc.) to receive the logs. I see it even on production firmware.

RTT does exactly what this PR does. It does buffer as long as there is space in the buffer and then either discards or blocks depending on configuration.

> I'd remove it. Or have it trigger only once. Both of these entirely fix the log loop problem, w.r.t. those logging statements, even if log output to CDC ACM is used.

OK, I will remove it.

tmon-nordic committed:
usb: device: cdc_acm: Always buffer poll out data if possible

Queue characters in TX ring buffer regardless of interface state. This
allows simple applications that print basic output to CDC ACM port to
not care about connection or DTR state. Note that depending on actual
application requirements it might be still wise to wait for DTR before
trying to send data.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
doc: connectivity: usb: Document CDC ACM ECHO mitigation

Explain the root cause behind garbage characters received in shell
applications, what Zephyr does to mitigate the issue and what is
expected from host userspace applications.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -1029,16 +1029,11 @@ static void cdc_acm_poll_out(const struct device *dev, unsigned char c)
 	while (!ring_buf_put(dev_data->tx_ringbuf, &c, 1)) {
 		if (k_is_in_isr() || !dev_data->flow_ctrl) {
-			LOG_INF("Ring buffer full, discard %c", c);
+			/* Ring buffer full, discard character. */
```
This LOG_INF must stay. If this bothers someone in the application, then USB_CDC_ACM_LOG_LEVEL should be disabled or the level set higher.

xyzzy42 replied:
You don't understand the implications of that.

xyzzy42 commented:
The new behavior here is a great improvement.  It's a shame the interrupt API doesn't also follow it too.

However, allowing the queue to fill, which is new behavior, and then generating log messages when it fills creates a log loop time bomb.

Someone is going to update working software to a new zephyr, and it's going to appear to work fine.  They'll be happy the console glitch in finally fixed.  But their device will randomly lock up or experience a massive slowdown.  Not at first.  Only after hours or even days of uptime when whatever logging they have on accumulates enough data to fill the ring buffer.  And that is not going to be a fun bug to track down!

Then they'll find this thread and see it was a known time bomb left in, and that they should have just known to turn off all CDC ACM logging (even though it was never triggering logging in earlier versions) and deserved what they got.  I don't think they'll be happy with that response.

fabiobaltieri commented:
> Then they'll find this thread and see it was a known time bomb left in, and that they should have just known to turn off all CDC ACM logging (even though it was never triggering logging in earlier versions) and deserved what they got. I don't think they'll be happy with that response.

Foreshadowing... would it be worth to add a compiler warning for this? Guess it could be captured with the right combo of macros.

tmon-nordic commented:
> > Then they'll find this thread and see it was a known time bomb left in, and that they should have just known to turn off all CDC ACM logging (even though it was never triggering logging in earlier versions) and deserved what they got. I don't think they'll be happy with that response.
> 
> Foreshadowing... would it be worth to add a compiler warning for this? Guess it could be captured with the right combo of macros.

I prefer to just remove the log entry. Let's move the discussion to #65112.

jfischer-no commented:
> The new behavior here is a great improvement. It's a shame the interrupt API doesn't also follow it too.
> 
> However, allowing the queue to fill, which is new behavior, and then generating log messages when it fills creates a log loop time bomb.

It will not create a log loop as you cannot enable logging in CDC ACM when it is used as a logging backend.

> Someone is going to update working software to a new zephyr, and it's going to appear to work fine. They'll be happy the console glitch in finally fixed. But their device will randomly lock up or experience a massive slowdown. Not at first. Only after hours or even days of uptime when whatever logging they have on accumulates enough data to fill the ring buffer. And that is not going to be a fun bug to track down!

Well you know for sure that "console glitch" is not Zephyr's problem and was only mitigated here, but it is funny why you still claim it.

xyzzy42 commented:
> It will not create a log loop as you cannot enable logging in CDC ACM when it is used as a logging backend.

I already explained earlier that this is not true.  But if you simply will not believe anything I say, just try it yourself.  Maybe Tomasz can explain it to you so you'll listen.

> Well you know for sure that "console glitch" is not Zephyr's problem and was only mitigated here, but it is funny why you still claim it.

I have no idea what you are trying to say here.  I'm pretty sure you are trying to insult me.  But it seems entirely unrelated to what you quoted.
